pyccel/stdlib/lists/lists.c

```diff
@@ -0,0 +1,665 @@
+
+        Parameters
+        ----------
+        list : t_list*
+            the t_list from where to remove the element
+
+        index : long int
+            index of the element to be removed
+
+        Returns
+        =======
+        ret_val : t_pop_ret
+*/
+{
+    size_t tsize = tSizes[list->type];
+    char *elements = (char*)(list->elements);
+    t_pop_ret *ret_val = malloc(sizeof(t_pop_ret));
+
+    index = (index < 0) ? list->size + index : index;
+    if (index >= 0 && list->size > index)
+    {
+        ret_val->type = list->type;
+        if (list->type == lst_list)
+            ret_val->raw = (char *)((t_list**)elements)[index];
+        else
+        {
+            ret_val->raw = malloc(list->type);
+            memcpy(ret_val->raw, &elements[index * tsize], tsize);
+        }
+        memmove(&elements[index * tsize], &elements[(index + 1) * tsize], list->size - index);
+        list->size -= 1;
+        return (ret_val);
+    }
+
+    return (NULL);
+}
+
+void    free_pop(t_pop_ret** pop_val)
+/*
+        Deallocate the t_pop_ret object resulting from a pop operation
+
+        Parameters
+        ----------
+        pop_val : t_pop_ret**
+            double pointer on the t_pop_ret object to be deallocated
+*/
+{
+    if (pop_val && (*pop_val))
+    {
+        free((*pop_val)->raw);
+        free(*pop_val);
+        *pop_val = NULL;
+    }
+}
+
+void     lst_remove(t_list* list, void* item)
+/*
+        removes an element by value
+
+        Parameters
+        ----------
+        list : t_list*
+            the t_list from where to remove the element
+
+        index : long int
+            the element to be removed
+*/
+{
+    int index;
+    t_pop_ret *ret;
+
+    if ((index = lst_index(list, item)) != -1)
+    {
+        ret = pop(list, (size_t)index);
+        free_pop(&ret);
+    }
+}
+
+void     reverse(t_list* list)
+/*
+        reverse the order of elements in a t_list
+
+        Parameters
+        ----------
+        list : t_list*
+            the t_list to be reversed
+*/
+{
+    size_t tsize = tSizes[list->type];
+    size_t index = 0;
+    char *elements = (char*)(list->elements);
+    void *buffer = malloc(tsize);
+
+    while (index + 1 <= list->size / 2)
+    {
+        memcpy(buffer, &elements[index * tsize], tsize);
+        memcpy(&elements[index * tsize], &elements[(list->size - index - 1) * tsize], tsize);
+        memcpy(&elements[(list->size - index - 1) * tsize], buffer, tsize);
+        index++;
+    }
+    free(buffer);
+}
+
+
+void*   array_subscripting(t_list *list, size_t index)
+/*
+        provides access to a specific element in a t_list
+
+        Parameters
+        ----------
+        list : t_list*
+            the t_list of the element to be accessed
+
+        index : long int
+            the element to be accessed
+*/
+{
+    char *elements = (char*)(list->elements);
+
+    if (list->size > index)
+        return ((void*)(&elements[index * tSizes[list->type]]));
+    return (NULL);
```

Comment:
Do you plan to include error checking in the code? If not then what happens when `NULL` is returned?

Reply:
Have you considered handling this with a macro as for ndarrays? Or with an inline function?

Reply:
Other than the size check is there a difference between this and `GET_INDEX`?

Reply:
yes error checking will be added, GET_INDEX and this function are the same
I think at some point @aihya added GET_INDEX to test his code.

Reply:
Has this been discussed? I thought there had been a discussion about error checking and we weren't sure we wanted it. I ought to have a discussion with Yaman about it at some point and see what his opinion is

Reply:
GET_INDEX is used to get an element from a list using an index, this function attempts to search for an element in the index.

Reply:
actually no I don't recall having a discussion about it